src/models/fire-model.ts

```diff
@@ -133,23 +129,57 @@ export const getFireSpreadRate = (sourceCell: Cell, targetCell: Cell, wind: IWin
           * Math.exp((0.792 + (0.681 * Math.pow(sav, 0.5)) ) * (packingRatio + 0.1));
 
-  const windSpeedFtPerMin = wind.speed * 88;
-  const windFactor = c * Math.pow(windSpeedFtPerMin, b) * Math.pow((packingRatio / optimumPackingRatio), -e);
-  const slopeFactor = 5.275 * Math.pow(packingRatio, -0.3) * Math.pow(Math.tan(slope), 2);
-
-  // Effective wind speed is defined in a bit strange way. It uses inverted wind factor equation to get effective
-  // wind speed using sum of wind factor and slope factor.
-  const effectiveWindSpeed = Math.pow((windFactor + slopeFactor) /
-          (c * Math.pow(packingRatio / optimumPackingRatio, -e)), 1 / b);
-
   const fuelLoad = netFuelLoad / (1 - totalMineralContent);
   const ovenDryBulkDensity = fuelLoad / fuelBedDepth;
 
   const effectiveHeatingNumber = Math.exp(-138 / sav);
 
   const heatOfPreIgnition = 250 + (1116 * moistureContent);
 
-  const directionFactor = getDirectionFactor(sourceCell, targetCell, effectiveWindSpeed, wind.direction);
-  return (reactionIntensity * propagatingFluxRatio * (1 + windFactor + slopeFactor)) * directionFactor
-    / (ovenDryBulkDensity * effectiveHeatingNumber * heatOfPreIgnition) / dist(sourceCell, targetCell);
+  // r0 is rate of spread without considering wind and slope.
+  const r0 = reactionIntensity * propagatingFluxRatio /
+             (ovenDryBulkDensity * effectiveHeatingNumber * heatOfPreIgnition);
+
+  const windSpeedFtPerMin = wind.speed * 88;
+  const windFactor = c * Math.pow(windSpeedFtPerMin, b) * Math.pow((packingRatio / optimumPackingRatio), -e);
+
+  const distInFt = dist(sourceCell, targetCell) * cellSize;
+  const elevationDiffInFt = targetCell.elevation - sourceCell.elevation;
+  const slopeTan = elevationDiffInFt / distInFt;
+  const slopeFactor = 5.275 * Math.pow(packingRatio, -0.3) * Math.pow(slopeTan, 2);
+
+  // Now, follow calculations from section "6.1 Fire Behavior in the Direction of Maximum Spread", p. 85.
+  // Note that "t" is assumed to be 1 to simplify calculations. Note that instead of assuming that slope vector
+  // is (ds, 0) and calculating angle between upslope vector and wind vector, we just use real angles, but set lengths
+  // correctly and use vector addition. It makes calculations more concise.
+
+  // 0 degrees is northern wind, so wind vector is pointing down (south). 90 deg should be eastern wind.
+  const windVector = (new Vector2(0, -1)).rotateAround(ORIGIN, -wind.direction * Math.PI / 180);
+  const upslopeVector = targetCell.elevation >= sourceCell.elevation ?
+          new Vector2(targetCell.x - sourceCell.x, targetCell.y - sourceCell.y) :
+          new Vector2(sourceCell.x - targetCell.x, sourceCell.y - targetCell.y);
+
+  const dw = r0 * windFactor; // * t (but t = 1)
+  windVector.setLength(dw);
+
+  const ds = r0 * slopeFactor; // * t (but t = 1)
+  upslopeVector.setLength(ds);
+
+  const maxSpreadRateVector = (new Vector2()).addVectors(upslopeVector, windVector);
```

Comment:
Oh, I see. Makes sense, I think, but all these factors laid on top of each other are making my head hurt!

Reply:
That's general comment re model / math, or some lines here? :wink:
I can try to simplify this code if you have some suggestions. Some constants are redundant, like `dw` and `ds`, but I did it, so it's easier to find references in Rothermel PDF.